src/core/math/quat.js

```diff
@@ -461,103 +476,63 @@ class Quat {
+        l = 1 / Math.sqrt(l);
+        m00 *= l;
+        m01 *= l;
+        m02 *= l;
+
+        l = m10 * m10 + m11 * m11 + m12 * m12;
+        if (l === 0) return this.set(0, 0, 0, 1);
+        l = 1 / Math.sqrt(l);
+        m10 *= l;
+        m11 *= l;
+        m12 *= l;
+
+        l = m20 * m20 + m21 * m21 + m22 * m22;
+        if (l === 0) return this.set(0, 0, 0, 1);
+        l = 1 / Math.sqrt(l);
+        m20 *= l;
+        m21 *= l;
+        m22 *= l;
+
+        // https://d3cw3dd2w32x2b.cloudfront.net/wp-content/uploads/2015/01/matrix-to-quat.pdf
+
+        if (m22 < 0) {
             if (m00 > m11) {
-                if (m00 > m22) {
-                    // XDiagDomMatrix
-                    rs = (m00 - (m11 + m22)) + 1;
-                    rs = Math.sqrt(rs);
-
-                    this.x = rs * 0.5;
-                    rs = 0.5 / rs;
-                    this.w = (m12 - m21) * rs;
-                    this.y = (m01 + m10) * rs;
-                    this.z = (m02 + m20) * rs;
-                } else {
-                    // ZDiagDomMatrix
-                    rs = (m22 - (m00 + m11)) + 1;
-                    rs = Math.sqrt(rs);
-
-                    this.z = rs * 0.5;
-                    rs = 0.5 / rs;
-                    this.w = (m01 - m10) * rs;
-                    this.x = (m20 + m02) * rs;
-                    this.y = (m21 + m12) * rs;
-                }
-            } else if (m11 > m22) {
-                // YDiagDomMatrix
-                rs = (m11 - (m22 + m00)) + 1;
-                rs = Math.sqrt(rs);
-
-                this.y = rs * 0.5;
-                rs = 0.5 / rs;
-                this.w = (m20 - m02) * rs;
-                this.z = (m12 + m21) * rs;
-                this.x = (m10 + m01) * rs;
+                this.set(1 + m00 - m11 - m22, m01 + m10, m20 + m02, m12 - m21);
             } else {
-                // ZDiagDomMatrix
-                rs = (m22 - (m00 + m11)) + 1;
-                rs = Math.sqrt(rs);
-
-                this.z = rs * 0.5;
-                rs = 0.5 / rs;
-                this.w = (m01 - m10) * rs;
-                this.x = (m20 + m02) * rs;
-                this.y = (m21 + m12) * rs;
+                this.set(m01 + m10, 1 - m00 + m11 - m22, m12 + m21, m20 - m02);
+            }
+        } else {
+            if (m00 < -m11) {
+                this.set(m20 + m02, m12 + m21, 1 - m00 - m11 + m22, m01 - m10);
+            } else {
+                this.set(m12 - m21, m20 - m02, m01 - m10, 1 + m00 + m11 + m22);
             }
         }
 
-        return this;
+        // Instead of scaling by 0.5 / Math.sqrt(t) (to match the original implementation),
+        // instead we normalize the result. It costs 3 more adds and muls, but we get
+        // a stable result and in some cases normalization is required anyway, see
+        // https://github.com/blender/blender/blob/v4.1.1/source/blender/blenlib/intern/math_rotation.c#L368
+        return this.mulScalar(1.0 / this.length());
```

Comment:
Isn't this just what `Quat#normalize` is, e.g. `return this.normalize()`? Then we also don't need to extend the API by adding a new method.

Reply:
Yep, similar, except we don't need to test for length 0. I wasn't sure the effort here was necessary, but we decided adding `mulScalar` to match the other types is probably a win.